RFC-0025-sdpa-optm-cpu.md

```diff
@@ -0,0 +1,79 @@
+# [PT2.1 Feature Proposal] SDPA (Scaled-Dot Product Attention) CPU Optimization
+
+This ticket is as part of PT 2.1 feature proposal process.
+
+## **Motivation**
+As LLM tends to accept a large batch size and a long context length, the requirement of large memory may lead to OOM issues or result in bad performance. To reduce memory usage and provide a substantial speedup for attention-related models, it is important to optimize SDPA. The fused SDPA, e.g. flash attention, is one type of the optimized SDPA algorithms designed for memory-bound problems, with better parallelism and memory access patterns. In PT 2.0, there exist both the basic unfused SDPA and the fused SDPA for CUDA, while only the unfused SDPA has CPU implementation. To fill the gap between CPU and CUDA, it is proposed to optimize SDPA by implementing the fused SDPA for CPU in PT 2.1.
+
+## **Implementation**
+We submitted PRs for CPU SDPA optimization and demonstrated up to 3x performance speedup on attention-related benchmarks.
+
+Here are the detailed implementation items:
+
+*   The flash attention CPU kernel is added, in which both forward and backward paths are implemented for data types float32 and bfloat16. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Specifically, FP32In-FP32Out and FP32In-BF16Out adopt the mkl gemm and BF16In-BF16Out adopts the OneDNN one. Parallelization is on the dimensions of batch size, head number and query length for forward path, and on the dimensions of batch size and head number for backward path. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.
```

Comment:
How is the mixed_dtype supposed to be handled ? Will it be built off of this: https://github.com/pytorch/pytorch/pull/103333?

Reply:
Currently, we only support the case where the three inputs, e.g. q, k, v, and the output own the same data type. Do you have any suggestions?

Reply:
This makes sense, I guess I was confused on `FP32In-BF16Out` but is this used for the second bmm after the softmax?

Reply:
Sorry, I accidently made a typo. Fixed from `FP32In-BF16Out` to `BF16In-FP32Out`, which is used for the two bmms in forward path when input dtype is BF16.